File: Projects/CPP-VNEXT/src/Position.cpp
#include "Position.h"

Position::Position(int x, int y): x(x), y(y) {}

int Position::GetX() const {
    return this->x;
}

int Position::GetY() const {
    return this->y;
}

Position Position::WithOffset(int dx, int dy) const {
    return Position(this->x + dx, this->y + dy);
}

File: Projects/Java/src/main/java/com/nieradko/worldsim/core/Position.java
package com.nieradko.worldsim.core;

import java.util.Objects;
import java.util.stream.Stream;

public abstract class Position {
    private final int x;
    private final int y;

    protected Position(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    boolean isWithinWorldOfSize(int n, int m) {
        return x >= 0
                && y >= 0
                && x < n
                && y < m;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj.getClass() != this.getClass()) {
            return false;
        }

        return ((Position)obj).x == x && ((Position)obj).y == y;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

    abstract Stream<Position> getAllNearbyPosition(int distance);
}
